zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,759 @@
+		block_count, allocation_bitarray_position);
+
+	/* Now that we cleared the allocation we just made we can check
+	 * if there was any other allocation in the space that would need to be
+	 * mapped. Since this API is only one that uses mapping in specified address
+	 * ranges we can assume that we made mapping for the space if it has any previous
+	 * allocations. We check that iterating over all physical allocation blocks
+	 */
+	physical_block_count = (phys_aligned_alloc_end
+		- phys_aligned_ptr) / CONFIG_MM_DRV_PAGE_SIZE;
+
+	for (i = 0; i < physical_block_count; i++) {
+		/* Check if beginning of physical block is misaligned with
+		 * allocation block size (ie block spans two physical pages)
+		 * Then we have to take one more allocation bit from beginning
+		 * to be sure that we did not miss page being already mapped
+		 * na - new allocation
+		 * allocator blocks [    ][    ][na11][1111][    ]
+		 * physical pages   [        ][mapped  ][mapped  ]
+		 * In example above we see that allocator block can span
+		 * two physical pages so we need to check overlapping in
+		 * both directions for logic to work.
+		 *
+		 * allocator [    ][    ][    ][x   ][    ]
+		 * physical  [        ][        ][        ]
+		 */
+		phys_block_ptr = phys_aligned_ptr + i * CONFIG_MM_DRV_PAGE_SIZE;
+		phys_block_end = phys_block_ptr + CONFIG_MM_DRV_PAGE_SIZE;
+
+		/* RFC do we need to align it down here ? Wouldn't it be automatically
+		 * aligned with both being powers of 2?
+		 */
+		block_realigned_ptr = ALIGN_DOWN(phys_block_ptr, block_size);
+		block_realigned_alloc_end = ALIGN_UP(phys_block_end, block_size);
+
+		check_offset = block_realigned_ptr
+			- (uintptr_t)heap->physical_blocks_allocators[mem_block_iterator]->buffer;
+		check_position = check_offset / block_size;
+
+		check_size = (block_realigned_alloc_end - block_realigned_ptr) / block_size;
+
+		/* We are now sure that if the allocation bit array between
+		 * block_realigned_ptr and block_realigned_alloc_end shows us if we already had
+		 * any allocations there and by our logic if it needs mapping or not and
+		 * we calculated position in bit array and size to check in bit array
+		 */
+		if (!sys_bitarray_is_region_cleared(
+				heap->physical_blocks_allocators[mem_block_iterator]->bitmap,
+				check_size, check_position))
+			continue;
+
+		/* needed in case of mid mapping failure
+		 * Rewrite of the concept is needed. Instead of failure handling we
+		 * need to check if there is enough physical memory available.
+		 */
+		if (!ptrs_to_map) {
+			ptrs_to_map = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED,
+				0, SOF_MEM_CAPS_RAM, sizeof(uintptr_t) * physical_block_count);
+			if (!ptrs_to_map)
+				goto fail;
+		}
+
+		ptrs_to_map[i] = phys_block_ptr;
+
+		if (sys_mm_drv_map_region((void *)phys_block_ptr, (uintptr_t)NULL,
+					CONFIG_MM_DRV_PAGE_SIZE, SYS_MM_MEM_PERM_RW) != 0) {
+			ptrs_to_map[i] = (uintptr_t)NULL;
+			goto fail;
+		}
```

Comment:
can we make a function `vmh_is_page_mapped()` and just call it for each physical page? Eventually for contiguous allocators you can just check the first and the last page, but that can be optimised later.

Reply:
I would refactor it after the implementation of zephyr physical memory counting - this will simplify this implementation and make it much shorter.